src/cms/models/users/user.py

```diff
@@ -0,0 +1,240 @@
+        verbose_name=_("last chat visit date"),
+        help_text=_("The date and time when the user did read the chat the last time"),
+    )
+    expert_mode = models.BooleanField(
+        default=False,
+        verbose_name=_("experienced user"),
+        help_text=_(
+            "Enable this option to show up additional features like XLIFF import/export, page filtering, mirrored pages, page-based permissions, Do-Not-Translate-Tag and recurring events"
+        ),
+    )
+
+    @property
+    def role(self):
+        """
+        We refer to Django user groups as roles.
+
+        :return: The role of this user
+        :rtype: ~cms.models.users.role.Role
+        """
+        if self.groups.exists():
+            return self.groups.first().role
+        return None
+
+    @property
+    def full_user_name(self):
+        """
+        Return the full name of the user. If either the first or the last name are present, return them, otherwise
+        return the username.
+
+        :return: The full name of the user
+        :rtype: str
+        """
+        return self.get_full_name() or self.get_username()
+
+    @property
+    def unread_chat_messages(self):
+        """
+        Return all unread messages of this user
+
+        :return: The unread messages of this user
+        :rtype: ~django.db.models.query.QuerySet [ ~cms.models.chat.chat_message.ChatMessage ]
+        """
+        return ChatMessage.history.filter(sent_datetime__gt=self.chat_last_visited)
+
+    def update_chat_last_visited(self):
+        """
+        Update the :attr:`~cms.models.users.user.User.chat_last_visited` to the current time
+
+        :return: Return the previous :attr:`~cms.models.users.user.User.chat_last_visited` value
+        :rtype: ~datetime.datetime
+        """
+        previous_chat_last_visited = self.chat_last_visited
+        self.chat_last_visited = timezone.now()
+        self.save()
+        logger.debug(
+            "Field chat_last_visited of %r updated from %s to %s",
+            self,
+            previous_chat_last_visited.strftime("%Y-%m-%d %H:%M:%S"),
+            self.chat_last_visited.strftime("%Y-%m-%d %H:%M:%S"),
+        )
+        return previous_chat_last_visited
+
+    def __str__(self):
+        """
+        This overwrites the default Django :meth:`~django.db.models.Model.__str__` method which would return ``User object (id)``.
+        It is used in the Django admin backend and as label for ModelChoiceFields.
+
+        :return: A readable string representation of the user
+        :rtype: str
+        """
+        return self.full_user_name
+
+    def __repr__(self):
+        """
+        This overwrites the default Django ``__repr__()`` method which would return ``<User: User object (id)>``.
+        It is used for logging.
+
+        :return: The canonical string representation of the user
+        :rtype: str
+        """
+        optional_fields = ""
+        if not self._state.adding:
```

Comment:
I could not reproduce this case. In which case is the user printed while it's in the adding process?
I think the fields `is_staff` etc. are present as soon as the object is initialized, so which errors can arise when the additional info is added when the user is in the `adding` state?

Reply:
True, this does not happen only in debug mode. I cannot remember the exact circumstances, but I had an error while creating the user. This resulted in an debug message being printed with user information. But as the error occurred before this was completed, it in turn created an error, because printing the debug message failed. This lead to an endless recursion and made finding the original problem complicated. Only after adding this, I was able to find the original problem.
 I'm fine with removing it, as this should usually not happen. On the other hand, it's probably not a problem to keep it, as the performance impact is probably negligible.

Reply:
I'd keep it as is right now, until anyone has or sees a problem with this.

Reply:
Ok, yeah I think it makes sense to make the output work even in the unlikely event that something goes wrong...
The only disadvantage I see is that `_state` is a private attribute, but since it's at least documented, I think it's not going anywhere soon...